### DISCUSSION TITLE
Do TIMESTEP_BEGIN and TIMESTEP_END make sense for t=0?

### DISCUSSION THREAD

**Discussion by felixcrazzolara** [idaholab/moose#28347] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I want to do a transient simulation of a set of beams. In order to verify that I set up the simulation correctly, I wanted to recover the expected resonance frequencies known from analytical solutions. To this end, I initialized 'disp_y' as given by the respective mode. For example, the initial position of the end of the beam, i.e. at 'x=10', has the initial value '-1'. This seems to work fine so far.

I also had to modify 'computeProperties' in ComputeIncrementalBeamStrain.C, since it wouldn't compute an incremental strain otherwise, but that shouldn't matter for my question.

Consider the four ways to log information about the displacement of the end of the beam:
```
 [disp_y_1]
   type = PointValue
   point = '10 0.0 0.0'
   variable = disp_y
   execute_on = "TIMESTEP_BEGIN"
 []
 [disp_y_2]
   type = PointValue
   point = '10 0.0 0.0'
   variable = disp_y
   execute_on = "TIMESTEP_END"
 []
 [disp_y_3]
   type = PointValue
   point = '10 0.0 0.0'
   variable = disp_y
   execute_on = "INITIAL TIMESTEP_BEGIN"
 []
[disp_y_4]
   type = PointValue
   point = '10 0.0 0.0'
   variable = disp_y
   execute_on = "INITIAL TIMESTEP_END"
 []
```

They only differ in how 'execute_on' is set. Below you can see the results for the first several timesteps:
```
t: [0.   0.01 0.02 0.03]
disp_y_1: [ 0.         -0.99999923 -0.81280427 -0.2029271 ]
disp_y_2: [ 0.         -0.81280427 -0.2029271   0.53472652]
disp_y_3: [-0.99999923 -0.99999923 -0.81280427 -0.2029271 ]
disp_y_4: [-0.99999923 -0.81280427 -0.2029271   0.53472652]
```
When I don't use 'INITIAL' in 'execute_on', the logged values are just zero. So my question is: What does it even mean TIMESTEP_BEGIN and TIMESTEP_END for t=0? How does it work under the hood?

Many thanks and best regards,

Felix

**Comment by GiudGiud** [comment]
Hello

INITIAL is an execution step that is right after the initialization of variables, and the solver has not run yet.
TIMESTEP_BEGIN is before the solver runs, and TIMESTEP_END after, as the solution is not the end-of-timestep solution

Both options makes sense, though timestep_end will not be at t=0 in your case, it will be at t=0.01

You can use the Debug/show_execution_order=ALWAYS to see the progression of the execution steps

**Comment by felixcrazzolara** [comment]
@GiudGiud thank you for your response! I think I didn't fully understand it yet. Let me ask you instead: In what case does it make sense to use TIMESTEP_END or TIMESTEP_BEGIN for t==0?

**Reply by GiudGiud** [reply]
It makes sense to use timestep begin if you are starting from t = 0 and want to output the initial solution. And same for the next time steps, you want the solution at the time which is at the beginning of the time step 

If you want the solution after the nonlinear solve, at the end of the time step (the beginning of the next one in a way, but objects may execute in between), then timestep_end is more appropriate

**Reply by felixcrazzolara** [reply]
Thank you very much again. I understand the second part you wrote, I think. The first part I don't.

The code snippet below shows how I set the initial conditions for the displacement in y-direction. From my initial post, it can be seen that the logged value of disp_y at x==10 is zero at t==0 when using 'execute_on=TIMESTEP_BEGIN' or 'execute_on=TIMESTEP_END'. For the lack of a better explanation, I'd say the logged output is "wrong" and hence useless. What am I missing? 
```
[Functions]
  [disp_y_ic_fun]
    type = ParsedFunction
    expression = '-0.050923532949828924*(sin(0.4694*x)-sinh(0.4694*x)) + 0.05000000020016891*(cos(0.4694*x)-cosh(0.4694*x))'
  []
[]

[ICs]
  [disp_y_ic]
    type = FunctionIC
    variable = disp_y
    function = disp_y_ic_fun
    block = 0
  []
[]
```

**Reply by GiudGiud** [reply]
Yoou are probably missing that you need both the postprocessors and the output to execute on INITIAL (or timestep_begin if you prefer) to see an updated result. 
Keep in mind, INITIAL is before time step begin. So many people use "INITIAL TIMESTEP_END" for the execute_on of any postprocessing (both the PPs and the outputs)
